### PR TITLE
Add support for iterating over batches in bfknn

### PR DESCRIPTION
**Pull request by benfred**
This adds support for iterating over batches of nearest neighbors in the brute force knn. This lets you query for the nearest neighbors, and then filter down the results - and if you have filtered out too many results, get the next batch of nearest neighbors.

The challenge here is balancing memory consumption versus efficieny: we could store the results of the full gemm in the distance calculation - but for large indices this discards the benefits of using the tiling strategy and risks running OOM. Instead we exponentially grow the number of neighbors being returned, and also cache both the query norms and index norms between calls.

### CODE REVIEW

**Review by wphicks — COMMENTED** [review]
I have not done an in-depth line-by-line review of this, but I wanted to provide feedback on the overall API. Short version is I _really_ like it! I was worried about storing the resources object at first, but the implementation both makes it clear why it is necessary and limits the opportunities for that to bite the user.

For downstream integrations, we're going to want to adopt some careful patterns for ensuring that we don't e.g. modify the index part way through batched searches, but this design will integrate _very_ cleanly with the work we're already doing in this direction.

This one gets a big +1 from me, and I can't wait to see similar capabilities for all types of indexes. Brilliant work!

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                           query,
+                           query_norms->view(),
+                           raft::linalg::NormType::L2Norm,
+                           raft::linalg::Apply::ALONG_ROWS);
+      }
+    }
+  }
+
+  /** a single batch of nearest neighbors in device memory */
+  class batch {
```
Comment (lowener) [inline cpp/include/raft/neighbors/brute_force_batch_k_query.cuh:106]: This `batch` class could be re-used for batching IVF and CAGRA searches. Can it be implemented in a separate file?

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+#include <raft/distance/distance_types.hpp>
+#include <raft/linalg/norm.cuh>
+#include <raft/matrix/slice.cuh>
+#include <raft/neighbors/brute_force.cuh>
+
+#include <raft/core/logger.hpp>
+
+namespace raft::neighbors::brute_force {
+/**
+ * @addtogroup brute_force
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force_batch_k_query.cuh]: good catch!

File: cpp/test/neighbors/tiled_knn.cu [hunk 3]
```diff
@@ -209,6 +213,72 @@ class TiledKNNTest : public ::testing::TestWithParam<TiledKNNInputs> {
...
+
+        offset += batch_size;
+        if (offset + batch_size > all_size) break;
+      }
+
+      // also test out with variable batch sizes
+      offset             = 0;
+      int64_t batch_size = k_;
+      batch_k_query<T, int> query(handle_, idx, query_view, batch_size);
+      for (auto it = query.begin(); it != query.end(); it.advance(batch_size)) {
```
Comment (cjnolet) [inline cpp/test/neighbors/tiled_knn.cu:254]: If the user is expected to invoke `advance()`, they should be able to pass in the `raft::resources` instance each time, right?

**Review by cjnolet — CHANGES_REQUESTED** [review]
Even though iterators are common in C++ collections APIs, this is a new design paradigm for RAFT and implementing brute-force first was a great idea to leave a small buffer for getting the API nice and tight. 

A couple of thoughts / intentional design decisions in RAFT
1. Using complex stateful objects is fine, but we should still facade them behind factory functions so that we're providing a layer of indirection between the object construction and the user. This gives us the ability to modify how the objects get constructed in the future without impacting the user. 
2. Trying to avoid storing `raft::resources` as state on underlying objects. This happened as a byproduct of the way we're constructing the container policies, but it was only dont with the intention of delaying the construction itself and isn't relied upon for any other state on the object. We're trying to move away from this pattern when we can avoid it. 
3. Exposing interfaces- by convention we use `*_types.hpp` to tell the user "these things are safe to expose on your own public APIs and they won't bring in additional dependencies. If RAFT is being compiled into a user's library, think of these as the sets of headers that are safe to expose in the users own `include/` directory. This means we should be hiding implementations for things like actual iterator types behidn `detail` as much as possible. It used to be okay to expose just cuda runtime things through `*_types.hpp` but we're trying to move away from that as well to support the potential that we might end up hosting Grace or other CPU-only APIs in the future. 

Most of my review comments are aesthetics, though, and I'm going to take a second pass over the implementation details but things look really good at a glance.

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  void load_batch(batch& output) const
+  {
+    std::optional<raft::device_vector_view<const float, int64_t>> query_norms_view;
+    if (query_norms) { query_norms_view = query_norms->view(); }
+
+    brute_force::search<T, IdxT>(
+      res, index, query, output.indices_.view(), output.distances_.view(), query_norms_view);
+  }
+
+  const raft::resources& res;
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force_batch_k_query.cuh:235]: This is also used through the `batch_k_query::begin()` and `batch_k_query::end()` methods.  If we want to support idiomatic C++ iterators (and by extension things like range based for loops), we will have to store the the `resources` object

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+      int64_t batch_size = std::min(std::max(offset * 2, next_batch_size * 2), query->index.size());
+      batches            = batch(query->res, query->query.extent(0), batch_size);
+      query->load_batch(batches);
+      current_batch_size = batch_size;
+    }
+
+    void slice_current_batch(int64_t offset, int64_t batch_size)
+    {
+      auto num_queries = batches.indices_.extent(0);
+      batch_size       = std::min(batch_size, query->index.size() - offset);
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force_batch_k_query.cuh:201]: The batch sizes can change as we iterate through K - if we want to support the redis `VecSimBatchIterator` interface, we will need this to handle the `getNextResults` method https://github.com/RedisAI/VectorSimilarity/blob/22954489d9184c9eba55f477463439a3532ca04e/src/VecSim/batch_iterator.h#L40-L42

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    iterator operator++(int)
+    {
+      iterator previous(*this);
+      operator++();
+      return previous;
+    }
+
+    void advance(int64_t next_batch_size)
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force_batch_k_query.cuh:172]: Yes - this method is for supporting variable batch sizes, which I think will be required for some of users.

**Comment by lowener** [comment]
Can you add some documentation and example to `docs/sourcs/vector_search_tutorial.md` Step 3? So that we can refer users trying to use batches to the tutorial.

**Comment by cjnolet** [comment]
> Can you add some documentation and example to docs/sourcs/vector_search_tutorial.md Step 3? So that we can refer users trying to use batches to the tutorial.

@lowener, I agree that this API should be included in the vector search tutorial, even if mentioned in the "additional features" section, but in the interest of time I think we should do a separate PR and target 24.02 with that change. We're just too close to burndown at this point.

File: cpp/include/raft/neighbors/brute_force-inl.cuh [hunk 7]
```diff
@@ -340,13 +340,15 @@ index<T> build(raft::resources const& res,
...
+ * @param[in] query_norms Optional device_vector_view of precomputed query norms
  */
 template <typename T, typename IdxT>
 void search(raft::resources const& res,
             const index<T>& idx,
             raft::device_matrix_view<const T, int64_t, row_major> queries,
             raft::device_matrix_view<IdxT, int64_t, row_major> neighbors,
-            raft::device_matrix_view<float, int64_t, row_major> distances)
+            raft::device_matrix_view<T, int64_t, row_major> distances,
+            std::optional<raft::device_vector_view<const T, int64_t>> query_norms = std::nullopt)
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force-inl.cuh]: I removed from the public api in the last commit (moving to detail )

File: cpp/include/raft/neighbors/brute_force_batch_k_query.cuh [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+   *
+   * Constructs the batch_k_query - which lets you iterate over batches of
+   * nearest neighbors.
+   *
+   * @param[in] res
+   * @param[in] index The index to query
+   * @param[in] query A device matrix view to query for [n_queries, index->dim()]
+   * @param[in] batch_size The size of each batch
+   */
+  batch_k_query(const raft::resources& res,
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/neighbors/brute_force_batch_k_query.cuh]: done in the last commit - the `batch_k_query` is now an abstract class containing no cuda code, and there is a `detail::gpu_batch_k_query` class with the cuda definitions that is created via a `make_batch_k_query` factory function

**Review by cjnolet — APPROVED** [review]

**Pull request merged**